Skeleton/src/Volunteer.cpp

```diff
@@ -1 +1,263 @@
+CollectorVolunteer* CollectorVolunteer::clone() const 
+{
+     return new CollectorVolunteer(*this);
+}
+void CollectorVolunteer::step() // need to check if it's correct
+{
+    if(decreaseCoolDown())
+    {
+        completedOrderId = activeOrderId;
+        activeOrderId = NO_ORDER;
+    }
+}
+int CollectorVolunteer::getCoolDown() const
+{
+    return coolDown;
+}
+
+int CollectorVolunteer::getTimeLeft() const 
+{
+    return timeLeft;
+}
+
+bool CollectorVolunteer::decreaseCoolDown() //Decrease timeLeft by 1,return true if timeLeft=0,false otherwise
+{
+    if(timeLeft > 0){
+        timeLeft = timeLeft - 1;
+    }
+    return (timeLeft == 0);
+}
+
+bool CollectorVolunteer::hasOrdersLeft() const  // Signal whether the volunteer didn't reach orders limit,Always true for CollectorVolunteer and DriverVolunteer
+{
+    return true;
+} 
+
+bool CollectorVolunteer::canTakeOrder(const Order &order) const 
+{
+    return (order.getStatus() == OrderStatus::PENDING && !Volunteer::isBusy());
+}
+
+void CollectorVolunteer::acceptOrder(const Order &order) 
+{
+    if(canTakeOrder(order)){
+        activeOrderId = order.getId();
+        timeLeft = coolDown;
+    }
+}
+string CollectorVolunteer::toString() const 
+{
+    return "Collector Volunteer - ID: " + std::to_string(getId())
+     + ", Name: " + getName() 
+     + ", Active Order ID: " + std::to_string(getActiveOrderId()) 
+     + ", Completed Order ID: " + std::to_string(getCompletedOrderId());
+}
+
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+//////////////////////////// LimitedCollector Volunteer //////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+
+LimitedCollectorVolunteer::LimitedCollectorVolunteer(int id, const string &name, int coolDown ,int maxOrders) : CollectorVolunteer(id, name, coolDown), maxOrders(maxOrders), ordersLeft(maxOrders) {}
+LimitedCollectorVolunteer* LimitedCollectorVolunteer::clone() const 
+{
+    return new LimitedCollectorVolunteer(*this);
+}
+
+bool LimitedCollectorVolunteer::hasOrdersLeft() const 
+{
+    return (ordersLeft > 0);
+}
+
+bool LimitedCollectorVolunteer::canTakeOrder(const Order &order) const 
+{
+    return (order.getStatus() == OrderStatus::PENDING && !CollectorVolunteer::isBusy() && hasOrdersLeft());
+}
+
+void LimitedCollectorVolunteer::acceptOrder(const Order &order) 
+{
+    if(canTakeOrder(order)){
+        ordersLeft--;
+        activeOrderId = order.getId();
+    }
+}
+
+int LimitedCollectorVolunteer::getMaxOrders() const 
+{
+    return maxOrders;
+}
+int LimitedCollectorVolunteer::getNumOrdersLeft() const
+{
+    return ordersLeft;
+}
+string LimitedCollectorVolunteer::toString() const 
+{
+        return "Limited Collector Volunteer - ID: " + std::to_string(getId()) +
```

Comment:
if you have access to private no need getter

Reply:
No access to private functions from a derived class, the binding will be in the 'Volunteer' class